src/i_3dsound.c

```diff
@@ -0,0 +1,312 @@
+{
+    src->position[0] = FIXED_TO_ALFLOAT(source->x);
+    src->position[2] = FIXED_TO_ALFLOAT(-source->y);
+
+    // Treat monsters and projectiles as point sources.
+    src->point_source = source->info && source->info->actualheight;
+
+    if (src->point_source)
+    {
+        const int yaw = source->angle >> ANGLETOFINESHIFT;
+
+        // Set vertical position to middle of sprite.
+        src->position[1] = FIXED_TO_ALFLOAT(source->z + (source->info->actualheight >> 1));
+
+        src->direction[0] = FIXED_TO_ALFLOAT(finecosine[yaw]);
+        src->direction[1] = 0.0f;
+        src->direction[2] = FIXED_TO_ALFLOAT(-finesine[yaw]);
+
+        src->velocity[0] = FIXED_TO_ALFLOAT(source->momx) * TICRATE;
+        src->velocity[1] = FIXED_TO_ALFLOAT(source->momz) * TICRATE;
+        src->velocity[2] = FIXED_TO_ALFLOAT(-source->momy) * TICRATE;
+    }
+    else
+    {
+        // This is a door, switch, lift, etc. and doesn't have a proper vertical
+        // position. Similar to vanilla Doom, make sure the player can hear this
+        // at any vertical position (e.g. tall lifts).
+        
+        // Set the source's vertical position to the listener's vertical position.
+        src->position[1] = FIXED_TO_ALFLOAT(listener->player->viewz);
+
+        src->direction[0] = 0.0f;
+        src->direction[1] = 0.0f;
+        src->direction[2] = 0.0f;
+
+        src->velocity[0] = 0.0f;
+        src->velocity[1] = 0.0f;
+        src->velocity[2] = 0.0f;
+    }
+}
+
+static void CalcHypotenuse(fixed_t adx, fixed_t ady, fixed_t *dist)
+{
+    if (ady > adx)
+    {
+        const fixed_t temp = adx;
+        adx = ady;
+        ady = temp;
+    }
+
+    if (adx)
+    {
+        const int slope = FixedDiv(ady, adx) >> DBITS;
+        const int angle = tantoangle[slope] >> ANGLETOFINESHIFT;
+        *dist = FixedDiv(adx, finecosine[angle]);
+    }
+    else
+    {
+        *dist = 0;
+    }
+}
+
+static void CalcDistance(const mobj_t *listener, const mobj_t *source,
+                         fixed_t *dist)
+{
+    const fixed_t adx = abs((listener->x >> FRACBITS) - (source->x >> FRACBITS));
+    const fixed_t ady = abs((listener->y >> FRACBITS) - (source->y >> FRACBITS));
+    const fixed_t adz = abs((listener->z >> FRACBITS) - (source->z >> FRACBITS));
+    fixed_t distxy;
+
+    CalcHypotenuse(adx, ady, &distxy);
+    CalcHypotenuse(distxy, adz, dist);
+}
+
+static boolean CalcVolumePriority(const mobj_t *listener, const mobj_t *source,
+                                  int *vol, int *pri)
+{
+    fixed_t dist;
+    int pri_volume;
+
+    CalcDistance(listener, source, &dist);
+
+    if (dist == 0)
+    {
+        return true;
+    }
+    else if (dist >= OAL_CULL_DISTANCE)
+    {
+        return false;
+    }
+    else if (dist <= OAL_REF_DISTANCE)
+    {
+        pri_volume = *vol;
+    }
+    else if (dist > OAL_MAX_DISTANCE)
+    {
+        // OpenAL inverse distance model never reaches zero volume. Gradually
```

Comment:
I don't quite get this part. If OpenAL has a volume-by-distance model, why can't we just use it at this point?

Reply:
An inverse model never reaches zero (e.g. `y = 1 / x`), so we have to ramp down to zero gradually. See [page 25](https://www.openal.org/documentation/openal-1.1-specification.pdf).

Reply:
That's understood, but why do we have to enforce reaching zero? In order to free the channel?

Reply:
I want to stay consistent with the custom priority system used by WinMBF which is based on volume. See `S_getChannel()` in `s_sound.c` and this same function in `i_mbfsound.c`.